walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,143 @@
+    boolean[] viableTimes = new boolean[TimeRange.END_OF_DAY+1];
+
+    // initially, all of the minutes are viable
+    for(int i=0;i<viableTimes.length;i++) {
+        viableTimes[i] = true;
+    }
+
+    // consider each event independently
+    iterator = events.iterator(); 
+    while(iterator.hasNext()) {
+        Event event = (Event) iterator.next();
+        Collection<String> eventAttendees = event.getAttendees();
+        TimeRange when = event.getWhen();
+        int eventStart = when.start();
+        int eventEnd = when.end();
+
+        // check if any of the attendees are attending the event
+        if(attending(attendeesMap, eventAttendees)) {
+            // mark all of the minutes from eventStart to eventEnd as false
+            for(int i=eventStart;i<eventEnd;i++) {
+                viableTimes[i] = false;
+            }
+        }
+    }
+    
+    return linearSearch(viableTimes, duration);
+  }
+
+  private static Collection<TimeRange> intersectionMeetingTimes(long duration, Collection<TimeRange> mandatoryViableMeetingTimes, Collection<TimeRange> optionalViableMeetingTimes) {
+    // all of the minutes when mandatory meetings can be held will be marked true
+    boolean[] mandatoryTimes = new boolean[TimeRange.END_OF_DAY+1];
+    Iterator iterator = mandatoryViableMeetingTimes.iterator(); 
+    while(iterator.hasNext()) {
+        TimeRange timeRange = (TimeRange) iterator.next();
+        int start = timeRange.start();
+        int end = timeRange.end();
+        
+        for(int i=start;i<end;i++) {
+            mandatoryTimes[i] = true;
+        }
+    }
+    
+    // all of the minutes when mandatory and optional meetings can be held will be marked true
+    boolean[] intersectionTimes = new boolean[TimeRange.END_OF_DAY+1];
+    iterator = optionalViableMeetingTimes.iterator(); 
+    while(iterator.hasNext()) {
+        TimeRange timeRange = (TimeRange) iterator.next();
+        int start = timeRange.start();
+        int end = timeRange.end();
+        
+        for(int i=start;i<end;i++) {
+            if(mandatoryTimes[i]) {
+                intersectionTimes[i] = true;
+            }
+        }
+    }
+
+    return linearSearch(intersectionTimes, duration);
+  }
+
+  private static boolean attending(HashMap<String, Boolean> attendeesMap, Collection<String> eventAttendees) {
+    Iterator iterator = eventAttendees.iterator();
+    while(iterator.hasNext()) {
+        if(attendeesMap.get(iterator.next()) != null)
+            return true;
+    }
+    return false;
+  }
+
+  /* Performs a linear search over the viable start times.
+   * For each potential start time, finds the largest contiguous viable time range beginning at that start time.
+   * If it's a viable meeting time (i.e. it lasts longer than the required duration), adds it to the output.
+  */
+  private static Collection<TimeRange> linearSearch(boolean[] viableTimes, long duration) {
```

Comment:
nit: Rename to something that captures the intent of the function. "linearSearch" is a little vague.